feat(mqtt sink): add MQTT sink

mladedav opened the pull request:
Continuation of #16065 which is itself a continuation of #13587

I've just fixed the intagration test and added a few changes as commented on in the last PR.

<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

mladedav commented:
@jszwedko It's ready as far as code is concerned. Everything else should be ready from the last PR if I understand the state correctly.

If there is something more to be done regarding CI I might need some guidance (e.g. failing spellchecker but it seems it is not required and the contributing guide does not mention it if I'm not overlooking it?)

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+				Whether or not end-to-end acknowledgements are enabled.
+
+				When enabled for a sink, any source connected to that sink, where the source supports
+				end-to-end acknowledgements as well, will wait for events to be acknowledged by the sink
```
```suggestion
				end-to-end acknowledgements as well, waits for events to be acknowledged by the sink
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+						This "encoding" simply uses the `message` field of a log event.
+
+						Users should take care if they're modifying their log events (such as by using a `remap`
+						transform, etc) and removing the message field while doing additional parsing on it, as this
```
```suggestion
						transform) and removing the message field while doing additional parsing on it, as this
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+						encoding that resembles the Prometheus export format.
+
+						Users should take care if they're modifying their log events (such as by using a `remap`
+						transform, etc) and removing the message field while doing additional parsing on it, as this
```
```suggestion
						transform) and removing the message field while doing additional parsing on it, as this
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+				}
+			}
+			except_fields: {
+				description: "List of fields that will be excluded from the encoded event."
```
```suggestion
				description: "List of fields to exclude from the encoded event."
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+				description: """
+					Controls how metric tag values are encoded.
+
+					When set to `single`, only the last non-bare value of tags will be displayed with the
```
```suggestion
					When set to `single`, only the last non-bare value of tags is displayed with the
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+					Controls how metric tag values are encoded.
+
+					When set to `single`, only the last non-bare value of tags will be displayed with the
+					metric.  When set to `full`, all metric tags will be exposed as separate assignments.
```
```suggestion
					metric.  When set to `full`, all metric tags are exposed as separate assignments.
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+				type: string: {
+					default: "single"
+					enum: {
+						full: "All tags will be exposed as arrays of either string or null values."
```
```suggestion
						full: "All tags are exposed as arrays of either string or null values."
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+					enum: {
+						full: "All tags will be exposed as arrays of either string or null values."
+						single: """
+															Tag values will be exposed as single strings, the same as they were before this config
```
```suggestion
															Tag values are exposed as single strings, the same as they were before this config
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+						full: "All tags will be exposed as arrays of either string or null values."
+						single: """
+															Tag values will be exposed as single strings, the same as they were before this config
+															option. Tags with multiple values will show the last assigned value, and null values will be
```
```suggestion
															option. Tags with multiple values show the last assigned value, and null values are
```

buraizu commented on website/cue/reference/components/sinks/base/mqtt.cue:
```diff
@@ -0,0 +1,290 @@
+				}
+			}
+			only_fields: {
+				description: "List of fields that will be included in the encoded event."
```
```suggestion
				description: "List of fields that are included in the encoded event."
```

buraizu reviewed (approved):
Approving with some minor edits suggested for `description` fields -- mainly around updating future tense to present

StephenWakely commented:
Amazing! Thank you so much. This is working beautifully now.

So the final minor things needed to get past the CI checks is the spelling. Can you add these words:

atleastonce
atmostonce
emqx
eventloop
exactlyonce
mqttoptions
rumqttc

to `.github/actions/spelling/allow.txt`.


We will also need a changelog, we just added this requirement after all the work was done previously on this sink so unfortunately it still needs doing. You can find details of how to add one [here]. [Here][example] is an example of one that was done for a new source.

We have a list [here][ci checks] of the various checks that you can run locally to ensure it will pass CI before we attempt to merge.

[here]: https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md
[example]: https://github.com/vectordotdev/vector/blob/master/changelog.d/10304_add_prometheus_pushgateway_source.feature.md
[ci checks]: https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#running-tests-locally

mladedav commented:
I've changed some stuff so that `mqttoptions`, `atleastonce` and so on were not used but were properly separated. I hope that's ok with the cue definitions.

I've added the changelog but didn't know what to write there except for the fact that it has been added. I've also added all people who have committed to this branch.

StephenWakely commented:
Looks like we have some Clippy errors too: https://github.com/vectordotdev/vector/actions/runs/7820956077/job/21357140481?pr=19813

mladedav commented:
I've fixed the changelog and fixed clippy. The compilation issue was introduced by me merging master into this branch through github ui, where something in the crate changed since this was written. 

Now everything should work, I just don't have cue locally so that test failed for me.

mladedav commented:
The failing test is `config::cmd::tests::output_has_consistent_ordering` and seems to shuffle around sink configurations and then checks that the output is the same.

I have purposefully made the configuration so that it creates random client ID so that there are no clashes between restarts or if the sink is used multiple times.

I've moved that later in the code so that no client ID means random client ID (the default) so that will hopefully fix the test.

neuronull commented:
👋  Hi @mladedav 

Thanks so much for taking this on.

I'll give the PR a proper review soon but I wanted to update you on the changelog / spelling stuff.

We had some changes to CI around the changelog yesterday and today. Your PR got caught in the cross fire.

Once this PR is merged: https://github.com/vectordotdev/vector/pull/19844

- merge master into your branch
- This commit should be reverted: https://github.com/vectordotdev/vector/pull/19813/commits/85d54ba9bd386dc5681751f0fcc8fc3e59ed479f
- The "authors" section in the changelog fragment, all the GH user names should not be prefixed with "@". With latest commits from master pulled in, this will now fail a CI check for that.

After that, the issues around spell checker and changelog fragments should be resolved. Thanks for your patience.

mladedav commented:
Thanks for the info, feel free to ping me if I don't notice it's been merged

neuronull commented:
> Thanks for the info, feel free to ping me if I don't notice it's been merged

👍  that has been merged @mladedav 

As an aside, it looks like your commits over the prior PR are only a couple, they make sense to me.

Since it has been a long time, I will do a scan over the entire changes to make sure there isn't anything we are missing due to the drift in the code base since the first and second attempts at this component.

neuronull commented on scripts/integration/mqtt/test.yaml:
```diff
@@ -0,0 +1,11 @@
+
+paths:
+- "src/internal_events/mqtt.rs"
+- "src/sinks/mqtt/**"
```
In general since all sinks use the `prelude.rs` which heavily leverages `util` module, we want to run all sink integration tests when that changes.

```suggestion
- "src/sinks/mqtt/**"
- "src/sinks/util/**"
```

neuronull commented on src/sinks/mqtt/config.rs:
```diff
@@ -0,0 +1,212 @@
+#[serde(rename_all = "lowercase")]
+#[allow(clippy::enum_variant_names)]
+pub enum MqttQoS {
+    /// AtLeastOnce.
+    #[derivative(Default)]
+    AtLeastOnce,
+
+    /// AtMostOnce.
+    AtMostOnce,
+
+    /// ExactlyOnce.
+    ExactlyOnce,
```
I noticed you changed the derivative default here. And I noticed in the rendered cue files we seem to not have a default, meaning the `quality_of_service` is a required field.  I'm wondering if the cue files need to be re-generated.

You can try `make check-component-docs`. to see if that is the case. If you don't already have CUE installed, this might not work for you. LMK and I can run this for you.

mladedav replied:
You're right. I've added a default function and now it is regenerated with the correct default.

neuronull reviewed (commented):
Looks good just a couple observations.

neuronull commented on src/sinks/mqtt/config.rs:
```diff
@@ -64,6 +64,7 @@ pub struct MqttSinkConfig {
     pub acknowledgements: AcknowledgementsConfig,
 
     #[configurable(derived)]
+    #[serde(default = "default_qos")]
```
🎉

neuronull reviewed (approved):
The new changes over the prior PRs look good.

I did a scan of the code in general, but I am approving this leaning on the thorough code reviews we did in the first two PRs.

StephenWakely reviewed (approved):
This is amazing! I'm really excited to be seeing this finally over the line.

jszwedko commented on scripts/integration/mqtt/test.yaml:
```diff
@@ -0,0 +1,12 @@
+features:
```
I think we need to run these tests to CI. That would mean adding to https://github.com/vectordotdev/vector/blob/master/.github/workflows/integration.yml . You can see the other tests there for an example.

neuronull replied:
good catch, added in https://github.com/vectordotdev/vector/pull/19813/commits/7dbb96edf54729e4a1bc08097885e3bf392acd54

neuronull commented:
This has merged into master now.

🎉 

Thanks to @mladedav and @zamazan4ik  for pushing through the remaining items on this sink.

Shout out to @astro , for your foundation of work on this component
